### PR TITLE
WebGLRenderer: Integrate PMREM, add setBackgroundBlurriness().

### LINKED ISSUE #20115: WebGLRenderer: Integrate PMREM
With the introduction of `WebGLCubeMaps`, the engine has now a mechanism to internally convert environment maps. I've quickly hacked in `PMREMGenerator` support today and it is actually possible to perform the conversion used in [webgl_loader_gltf](https://threejs.org/examples/webgl_loader_gltf). It would love to see PMREM integrated in the renderer since it's a technical detail most users should not care about.

However, my implementation is a hack right now since it's not defined _when_ `PMREMGenerator` should do its work. The question is: How can `WebGLCubeMaps` detect textures that should be processed by `PMREMGenerator`?

### PR DESCRIPTION
**Pull request by Mugen87**
Related issues:

Fixed #20115.

**Description**

This PR integrates `PMREMGenerator` such that cube maps and equirectangular environment maps are automatically converted to the cubeUV format for PBR materials.

Besides, `WebGLRenderer.setBackgroundBlurriness()` is added which allows to control the blurriness of skyboxes. This method also works when a cube or  equirectangular texture is assigned to `Scene.background` by converting these texture into the cubeUV format. It essentially addresses my feedback from https://github.com/mrdoob/three.js/pull/18241#issuecomment-569432416.

`webgl_loader_gltf` was the only example which was updated in this PR. The usage of `WebGLRenderer.setBackgroundBlurriness()` is important in this example otherwise the background would look very blocky (because of the nearest filtered HDR texture). I have applied a very blurred background for now (which you sometimes see in other engines):

image

### CODE REVIEW

**Comment by Mugen87** [comment]
@mrdoob I hope we can manage to review this PR in the `r123` dev cycle. It would definitely simplify the creation of PBR setups if user don't have to work with `PMREMGenerator` anymore.

**Comment by Mugen87** [comment]
@mrdoob If you have concerns about `WebGLRenderer.setBackgroundBlurriness()`, it should be no problem to replace it later with your suggested API in #17420. This PR is mainly focused on integrating `PMREMGenerator` into the renderer. However, a control mechanism for a skybox's blurriness is required. Hence the addition of `WebGLRenderer.setBackgroundBlurriness()`.

**Comment by Mugen87** [comment]
I have encountered an issue during testing today. There is a problem when using a cube render target as an environment map for a PBR material. In this case, `WebGLCubeMaps` tries to convert it to cubeUV format which fails currently because `image.length === 6` is not true for cube render targets.

Besides, `WebGLCubeMaps` performs its conversions just once and not every time the render target is updated e.g. when using `CubeCamera`. On the other hand, it's not recommended to apply cube maps generated by `CubeCamera` to a PBR material since they are not pre-processed with `PMREMGenerator`. Doing this per frame is surely not good for performance anyway.

**Comment by Mugen87** [comment]
I have noticed this in [webgl_materials_envmaps_parallax](https://threejs.org/examples/webgl_materials_envmaps_parallax). I think this example is not correct since it applies a non `PMREMGenerator` preprocessed env map to `MeshPhysicalMaterial`.

**Comment by Mugen87** [comment]
Okay, I have found an easy solution. `WebGLCubeMaps` only processes textures with a version greater zero now which is not true for render targets. If users work with render targets and need a conversion to cubeUV, they can handle it by themselves via `PMREMGenerator`. In this case, they have better control on when to regenerate the cubeUV env map again (which is not required for static textures).

**Comment by Mugen87** [comment]
Um, it seems the only reliable way to make this more robust is to introduce some sort of flag or subclass that allows us to distinct textures of render targets. In this way, `WebGLCubeMaps` could easily detect textures of render targets and exclude any automatic pre-processing of environment maps.

**Comment by mrdoob** [comment]
Can we leave the background blurriness for another PR? I think this PR should focus on integrating PMREM.

**Comment by Mugen87** [comment]
The current implementation uses the material to determine whether `PMREMGenerator` should be used or not. Meaning when using a PBR material with an equirectangular or cube map,  `WebGLCubeMaps` performs the conversion to the cubeUV format.

However, this workflow does not work for `WebGLBackground` since the skybox is rendered with a special shader material. And skyboxes should work no matter if you have meshes with PBR materials in your scene. 

So if there is no way to control the blurriness, the renderer does not know whether to blur the skybox or not. This is required though otherwise it is not possible to replicate the current looks of the examples. Meaning if the feature is missing, the background will lock blocky because of the nearest filtered HDR textures.

The only workaround would be to ask the user to use `PMREMGenerator` on application level in order to produce a cubeUV texture for `Scene.background`. However, this would be a bit contradictory to the actual goal of the PR: The avoidance of `PMREMGenerator` on application level.

The issue becomes more clear if you download the branch and remove the line `renderer.setBackgroundBlurriness( 0.5 );` from `webgl_loader_gltf`. The skybox will look blocky. I'm not sure how a different implementation should look like so `WebGLBackground` always requests the correct env map from `WebGLCubeMaps`.

**Comment by mrdoob** [comment]
Hmm, okay.
I need to give it a go myself so I have a good understanding of the problem.
Will use this PR as reference.

**Comment by mrdoob** [comment]
Hmm, could you try splitting `WebGLCubeMaps` into `WebGLCubeMaps` and `WebGLCubeUVMaps`?

**Comment by Mugen87** [comment]
I've tried this before but it means we need at each place where `cubemaps.get()` is now called additional (and thus redundant) code. Because it's necessary to distinct whether `WebGLCubeMaps` or `WebGLCubeUVMaps` should be used. 

If we want to go this route, we probably need an additional component that handles this. Maybe `WebGLEnvMaps`?

**Comment by mrdoob** [comment]
> I've tried this before but it means we need at each place where `cubemaps.get()` is now called additional (and thus redundant) code. Because it's necessary to distinct whether `WebGLCubeMaps` or `WebGLCubeUVMaps` should be used. 

Wouldn't moving the logic outside be something like this?

```js
materialProperties.envMap = ( material.isMeshStandardMaterial ? cubeuvmaps : cubemap ).get( material.envMap || environment );
```

**Comment by Mugen87** [comment]
Um...I'm not sure anymore to be honest 😅 . I would have to give this a try again^^.

**Comment by Mugen87** [comment]
Sorry, I have to dive into this topic again. There are different ways how `cubemaps.get()` or `cubeuvmaps.get()` could be called and I have to verify all of them.

**Comment by Mugen87** [comment]
> Hmm, could you try splitting WebGLCubeMaps into WebGLCubeMaps and WebGLCubeUVMaps?

Okay, this worked fine! But it still necessary to solve the render target issue...

**Comment by Mugen87** [comment]
I've introduced `isRenderTargetTexture` to detect textures which are owned by render targets. In this way, it's possible to not break existing code and make auto-conversion more robust. It makes also `CubeTexture._needsFlipEnvMap` obsolete.

The policy is that the engine does not auto-convert environment maps if they are textures of render targets. If this case, the user have to do it if desired.

**Comment by WestLangley** [comment]
/bump

I support both the integration of PMREM and background blurriness.

However, I really would like to have the `MeshStandardMaterial` and `MeshPhysicalMaterial` shaders only support `CUBE_UV` environment maps, because only `CUBE_UV` maps work correctly.

So... supporting the auto-conversion of any user-specified `Standard/Physical` environment map to PMREM would be ideal.

We could then remove PMREM from the API -- or keep it as a /jsm utility if users want to create there own PMREMs.

@Mugen87 Is that doable at this point, or are there still "gotchas" to deal with?

**Comment by Mugen87** [comment]
> However, I really would like to have the MeshStandardMaterial and MeshPhysicalMaterial shaders only support CUBE_UV environment maps, because only CUBE_UV maps work correctly.

Makes sense. But since it's a breaking change and for simplicity reasons, I would make a different PR for this.

> We could then remove PMREM from the API -- or keep it as a /jsm utility if users want to create there own PMREMs.

I suggest to keep it in order to not break existing apps. At some point in the future it's maybe okay to make the class internal.

> Is that doable at this point, or are there still "gotchas" to deal with?

I think the PR was considered to be complete. However, if we agree to an API I would make a fresh one since the existing merge conflicts are already severe.

**Comment by WestLangley** [comment]
>But since it's a breaking change

What would break?

**Comment by Mugen87** [comment]
Sorry, I've mixed topics. After studying the code again cube uv is in fact the only supported format since the other two (equirect and cubemap) are converted anyway. So it's just a matter of cleaning up the obsolete code.

**Comment by WestLangley** [comment]
>So it's just a matter of cleaning up the obsolete code.

... in the shader. And I'll do that as a second step if you can automate the computation of PMREM. 😇

**Comment by WestLangley** [comment]
@Mugen87

And to clarify, equirect is converted to cube map currently, but for PBR materials, it will have to be converted directly to PMREM, instead.

Cubemap will also be converted directly to PMREM.

**Comment by Mugen87** [comment]
> And to clarify, equirect is converted to cube map currently, but for PBR materials, it will have to be converted directly to PMREM, instead.

Correct! This is the goal of this PR!

**Comment by Mugen87** [comment]
Closing in favor of #22178.

File: src/renderers/webgl/WebGLCubeUVMaps.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+				if ( cubeUVmaps.has( texture ) ) {
+
+					return cubeUVmaps.get( texture ).texture;
+
+				} else {
+
+					const image = texture.image;
+
+					if ( ( isEquirectMap && image && image.height > 0 ) || ( isCubeMap && image && image.length === 6 ) ) {
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLCubeUVMaps.js:31]:
```js
( isCubeMap && image && image.length === 6 )
```
This line was actually buggy. `image.length === 6` is not safe for testing if cube textures are complete. The cube texture loader doesn't load the single images in order so the array can have a length of six although images are still missing.

**Pull request closed without merging**